tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -3243,6 +3299,39 @@ port::Status CudnnSupport::DoBatchNormalizationForwardImpl(
   float zero = 0.0;
   auto cudnn = cudnn_->GetHandle(parent_, stream);
 
+#if CUDNN_VERSION >= 7402
+  cudnnBatchNormOps_t bn_ops = CUDNN_BATCHNORM_OPS_BN;
+#endif
+  DeviceMemory<uint8> workspace;
+  DeviceMemory<uint8> reserve_space;
+  if (reserve_space_allocator != nullptr) {
+#if CUDNN_VERSION >= 7402
+    SE_ASSIGN_OR_RETURN(workspace,
+                        CreateBatchNormForwardWorkspace(
+                            stream, cudnn, mode, bn_ops, x_descriptor,
+                            scale_offset_descriptor, workspace_allocator))
+    if (is_training) {
+      size_t reserve_space_size_in_bytes = 0;
+      RETURN_IF_CUDNN_ERROR(
+          cudnnGetBatchNormalizationTrainingExReserveSpaceSize(
+              /*handle=*/cudnn.handle(), /*mode=*/mode, /*bnOps=*/bn_ops,
+              /*activationDesc=*/NULL, /*xDesc=*/x_descriptor.handle(),
+              /*sizeInBytes=*/&reserve_space_size_in_bytes));
+      SE_ASSIGN_OR_RETURN(reserve_space,
+                          reserve_space_allocator->AllocateBytes(
+                              stream, reserve_space_size_in_bytes));
+    }
+#else
+    if (is_training) {
+      // dummy allocation for reserve_space
+      size_t reserve_space_size_in_bytes = 0;
+      SE_ASSIGN_OR_RETURN(reserve_space,
```

Comment:
It seems to me that `reserve_space` is not used below in this case, then why do we still allocate it?

Reply:
This branch `if (is_training)` is in the branch `if (reserve_space_allocator != nullptr)`, meaning the V3 op is used and thus this reserve_space should exist. As you mentioned, if cudnn < 7.4.2, we will fall back to use older cudnn api and don't actually use this reserve_space, but in the graph, the op is still called V3 and we need to handle it as a dummy parameter.